hilti/toolchain/src/ast/expression.cc

```diff
@@ -1,3 +1,15 @@
 // Copyright (c) 2020-2021 by the Zeek Project. See LICENSE for details.
 
 #include <hilti/ast/expression.h>
+#include <hilti/ast/expressions/id.h>
+
+using namespace hilti;
+
+bool expression::isResolved(const detail::Expression& e, type::ResolvedState* rstate) {
+    // We always consider `self` expressions as fully resolved to break the
+    // cycle with the type that they are pointing to.
+    if ( auto id = e.tryAs<expression::ResolvedID>(); id && id->id() == ID("self") )
```

Comment:
With this any identifier `self` will always be considered resolved, right? We currently allow this code and it does what is expected:

```ruby
module foo;
global self = 42;
print self;
```

Even with this patch this example still resolves successfully and behaves the same, but I wonder whether there are other cases which would now be broken.

Maybe we could make this less of an issue if we prevented users from creating their own bindings named `self` (could think of variable and parameter decls, not sure there is more).

Reply:
I guess this works since in this example `self` would always have an id `foo::self`. I also tried the following:

```ruby

module foo;

function x() : uint64 {
	local self = 42;
	return self;
}
print x();
```
This still works even though `self` now has an id `self` 🤯.

Reply:
I'm adding a general check that declarations can't redefine `self`.